### PR TITLE
mem: use slice capacity instead of length, to determine whether to pool buffers or directly allocate them

### LINKED ISSUE #7631: mem.NewBuffer() should look at slice capacity?
### What version of gRPC are you using?

1.66.2

### What version of Go are you using (`go version`)?

N/A

### What operating system (Linux, Windows, …) and version?

N/A

### What did you do?

Roughly this:

```go
buf := pool.Get(32*1024)

_, _ = reader.Read(buf)

buffer := mem.NewBuffer(buf, pool)

buffer.Free()
```

I'm getting a ~big buffer from the pool and reading into it.

### What did you expect to see?

I expect to get a "normal" buffer for the `buf` slice. That way it can be put back into the pool when it's no longer needed.

### What did you see instead?

`mem.NewBuffer()` looks at the length of the `buf` and, some times, it's under the threshold. In that case the `buf` is wrapped into `SliceBuffer` and returned as is. This means `buf` will not be returned into the pool, but become garbage. This defeats the point of buffer pooling, obviously.

Perhaps `mem.NewBuffer()` should look at slice capacity (`cap(buf)`) instead of length (`len(buf)`)?

---

Current (undocumented) behavior contradicts the function documentation too:

```
// NewBuffer creates a new Buffer from the given data, initializing the reference
// counter to 1. The data will then be returned to the given pool when all
// references to the returned Buffer are released. As a special case to avoid
// additional allocations, if the given buffer pool is nil, the returned buffer
// will be a "no-op" Buffer where invoking Buffer.Free() does nothing and the
// underlying data is never freed.
//
// Note that the backing array of the given data is not copied.
```

The buffer will not be returned to the pool if it's length is too small.

### PR DESCRIPTION
**Pull request by PapaCharlie**
Fixes #7631 

As the issue states, `mem.NewBuffer` would not pool buffers with a length below the pooling threshold but whose capacity is actually larger than the pooling threshold. This can lead to buffers being leaked.

RELEASE NOTES:
- mem: use slice capacity instead of length, to determine whether to pool buffers or directly allocate them

### CODE REVIEW

**Comment by purnesh42H** [comment]
@dfawley @easwars for second review

File: mem/buffers.go
Comment (easwars) [inline mem/buffers.go:97]: Nit: Can you please wrap the comment lines, here and elsewhere in this PR, to 80-cols. Thanks.

File: mem/buffers_test.go
Comment (easwars) [inline mem/buffers_test.go:123]: Nit: s/buffer/Buffer

**Review by easwars — APPROVED** [review]
LGTM, modulo minor nits.

**Pull request merged**
